crypto/ec/ec2_lambda.c

```diff
@@ -0,0 +1,442 @@
+    else if (a->Z_is_one)
+        ret = ec_GF2m_lambda_add_mixed(group,r,b,a,ctx);
+    else
+        ret = ec_GF2m_lambda_add_proj(group,r,a,b,ctx);
+
+    r->Z_is_one = 0;
+    return ret;
+}
+
+/**
+ * Inverts point wrt lambda-affine/proj coords.
+ * I.e. -(X, Y, Z) -> (X, Y XOR Z, Z)
+ *
+ * @param point input and output point
+ */
+int ec_GF2m_lambda_invert(const EC_GROUP *group, EC_POINT *point, BN_CTX *ctx)
+{
+    /* -infty -> infty */
+    if (EC_POINT_is_at_infinity(group, point))
+        return 1;
+
+    return BN_GF2m_add(point->Y, point->Y, point->Z);
+}
+
+/**
+ * Sets point to lambda-affine coord equiv of short coords (x,y)
+ * I.e. (x, y) -> (x, x XOR y/x, 1)
+ * @param point output point in lambda-affine
+ * @param x input coord in short-affine
+ * @param y input coord in short-affine
+ */
+int ec_GF2m_lambda_point_set_affine_coordinates(const EC_GROUP *group,
+                                                EC_POINT *point,
+                                                const BIGNUM *x,
+                                                const BIGNUM *y, BN_CTX *ctx)
+{
+    int ret = 0;
+
+    if (x == NULL || y == NULL) {
+        ECerr(EC_F_EC_GF2M_LAMBDA_POINT_SET_AFFINE_COORDINATES,
+              ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+
+    if (!BN_copy(point->X, x))
+        goto done;
+    BN_set_negative(point->X, 0);
+    if (!group->meth->field_div(group, point->Y, y, point->X, ctx)
+            || !BN_GF2m_add(point->Y, point->Y, point->X))
+        goto done;
+    BN_set_negative(point->Y, 0);
+    if (!BN_one(point->Z))
+        goto done;
+    BN_set_negative(point->Z, 0);
+    point->Z_is_one = 1;
+    ret = 1;
+
+ done:
+    return ret;
+}
+
+/**
+ * Sets x and y to short-affine equiv of lambda-affine/proj point.
+ * I.e. (X, Y, Z) -> (X/Z, Y/Z, 1) then
+ * (x, y) -> (x, (x XOR y) * x)
+ *
+ * @param point input point in lambda-affine/proj
+ * @param x output coord in short-affine. Can be NULL.
+ * @param y output coord in short-affine. Can be NULL.
+ */
+int ec_GF2m_lambda_point_get_affine_coordinates(const EC_GROUP *group,
+                                                const EC_POINT *point,
+                                                BIGNUM *x, BIGNUM *y,
+                                                BN_CTX *ctx)
+{
+    int ret = 0;
+    EC_POINT *r;
+
+    if (EC_POINT_is_at_infinity(group, point)) {
+        ECerr(EC_F_EC_GF2M_LAMBDA_POINT_GET_AFFINE_COORDINATES,
+              EC_R_POINT_AT_INFINITY);
+        return 0;
+    }
+
+    r = EC_POINT_new(group);
+    if (r == NULL
+            || !EC_POINT_copy(r, point)
+            || !EC_POINT_make_affine(group, r, ctx))
+        goto done;
+    if (x != NULL) {
+        if (!BN_copy(x, r->X))
+            goto done;
+        BN_set_negative(x, 0);
+    }
+    if (y != NULL) {
+        if (!BN_GF2m_add(y, r->Y, r->X)
+                || !group->meth->field_mul(group, y, y, point->X, ctx))
```

Comment:
@bbbrumley Should this be `r->X` here?

Reply:
Making that change seems to fix most tests except for one, for this curve:

    /* IPSec curves */
    {NID_ipsec3, &_EC_IPSEC_155_ID3.h, 0,
     "\n\tIPSec/IKE/Oakley curve #3 over a 155 bit binary field.\n"
     "\tNot suitable for ECDSA.\n\tQuestionable extension field!"},